File: test/domain/agent/risk_classifier_test.dart
import 'package:aios/domain/agent/risk_classifier.dart';
import 'package:aios/domain/entities/agent_models.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  late RiskClassifier classifier;

  setUp(() {
    classifier = RiskClassifier();
  });

  group('classifyRisk_safeTools', () {
    test('classify_calculator_returnsSafe', () {
      expect(classifier.classify('calculator', '{}'), ToolRisk.safe);
    });

    test('classify_timer_returnsSafe', () {
      expect(classifier.classify('timer', '{}'), ToolRisk.safe);
    });

    test('classify_deviceInfo_returnsSafe', () {
      expect(classifier.classify('device_info', '{}'), ToolRisk.safe);
    });

    test('classify_notepad_returnsSafe', () {
      expect(classifier.classify('notepad', '{}'), ToolRisk.safe);
    });

    test('classify_screenReader_returnsSafe', () {
      expect(classifier.classify('screen_reader', '{}'), ToolRisk.safe);
    });

    test('classify_screenFind_returnsSafe', () {
      expect(classifier.classify('screen_find', '{}'), ToolRisk.safe);
    });

    test('classify_notificationReader_returnsSafe', () {
      expect(
        classifier.classify('notification_reader', '{}'),
        ToolRisk.safe,
      );
    });

    test('classify_contactSearch_returnsSafe', () {
      expect(classifier.classify('contact_search', '{}'), ToolRisk.safe);
    });
  });

  group('classifyRisk_smsSender', () {
    test('classify_smsSenderSend_returnsCritical', () {
      expect(
        classifier.classify('sms_sender', '{"action": "send"}'),
        ToolRisk.critical,
      );
    });

    test('classify_smsSenderRead_returnsHigh', () {
      expect(
        classifier.classify('sms_sender', '{"action": "read"}'),
        ToolRisk.high,
      );
    });
  });

  group('classifyRisk_phoneCaller', () {
    test('classify_phoneCallerCall_returnsCritical', () {
      expect(
        classifier.classify('phone_caller', '{"action": "call"}'),
        ToolRisk.critical,
      );
    });

    test('classify_phoneCallerDial_returnsHigh', () {
      expect(
        classifier.classify('phone_caller', '{"action": "dial"}'),
        ToolRisk.high,
      );
    });
  });

  group('classifyRisk_screenAction', () {
    test('classify_screenActionTap_returnsHigh', () {
      expect(
        classifier.classify('screen_action', '{"action": "tap"}'),
        ToolRisk.high,
      );
    });

    test('classify_screenActionGlobal_returnsLow', () {
      expect(
        classifier.classify('screen_action', '{"action": "global"}'),
        ToolRisk.low,
      );
    });

    test('classify_screenActionTypePassword_returnsCritical', () {
      expect(
        classifier.classify(
          'screen_action',
          '{"action": "type", "content": "enter your password"}',
        ),
        ToolRisk.critical,
      );
    });

    test('classify_screenActionTypePin_returnsCritical', () {
      expect(
        classifier.classify(
          'screen_action',
          '{"action": "type", "content": "enter your pin"}',
        ),
        ToolRisk.critical,
      );
    });

    test('classify_screenActionTypeCvv_returnsCritical', () {
      expect(
        classifier.classify(
          'screen_action',
          '{"action": "type", "content": "cvv code"}',
        ),
        ToolRisk.critical,
      );
    });

    test('classify_screenActionTypeOtp_returnsCritical', () {
      expect(
        classifier.classify(
          'screen_action',
          '{"action": "type", "content": "otp verification"}',
        ),
        ToolRisk.critical,
      );
    });

    test('classify_screenActionTypeNormal_returnsHigh', () {
      expect(
        classifier.classify(
          'screen_action',
          '{"action": "type", "content": "hello world"}',
        ),
        ToolRisk.high,
      );
    });
  });

  group('classifyRisk_appLauncher', () {
    test('classify_appLauncherOpenApp_returnsHigh', () {
      expect(
        classifier.classify('app_launcher', '{"action": "open_app"}'),
        ToolRisk.high,
      );
    });

    test('classify_appLauncherOpenSettings_returnsLow', () {
      expect(
        classifier.classify('app_launcher', '{"action": "open_settings"}'),
        ToolRisk.low,
      );
    });
  });

  group('classifyRisk_edgeCases', () {
    test('classify_unknownTool_returnsHigh', () {
      expect(
        classifier.classify('nonexistent_tool', '{}'),
        ToolRisk.high,
      );
    });

    test('classify_invalidJson_defaultsToEmptyAction', () {
      expect(classifier.classify('calculator', 'not json'), ToolRisk.safe);
      expect(
        classifier.classify('screen_action', 'not json'),
        ToolRisk.high,
      );
      expect(
        classifier.classify('app_launcher', 'not json'),
        ToolRisk.low,
      );
    });

    test('classify_caseInsensitiveAction_classifiesCorrectly', () {
      expect(
        classifier.classify('sms_sender', '{"action": "SEND"}'),
        ToolRisk.critical,
      );
      expect(
        classifier.classify('phone_caller', '{"action": "CALL"}'),
        ToolRisk.critical,
      );
    });

    test('classify_emptyArgs_defaultsAppropriately', () {
      expect(classifier.classify('calculator', ''), ToolRisk.safe);
      expect(classifier.classify('screen_action', ''), ToolRisk.high);
    });

    test('classify_appLauncherListApps_returnsLow', () {
      expect(
        classifier.classify('app_launcher', '{"action": "list_apps"}'),
        ToolRisk.low,
      );
    });

    test('classify_appLauncherOpenUrl_returnsHigh', () {
      expect(
        classifier.classify('app_launcher', '{"action": "open_url"}'),
        ToolRisk.high,
      );
    });

    test('classify_screenActionTypeSsn_returnsCritical', () {
      expect(
        classifier.classify(
          'screen_action',
          '{"action": "type", "content": "enter your ssn"}',
        ),
        ToolRisk.critical,
      );
    });

    test('classify_screenActionTypePasscode_returnsCritical', () {
      expect(
        classifier.classify(
          'screen_action',
          '{"action": "type", "content": "passcode"}',
        ),
        ToolRisk.critical,
      );
    });

    test('classify_screenActionTypeSocialSecurity_returnsCritical', () {
      expect(
        classifier.classify(
          'screen_action',
          '{"action": "type", "content": "enter your social security number"}',
        ),
        ToolRisk.critical,
      );
    });

    test('classify_screenActionLongClick_returnsHigh', () {
      expect(
        classifier.classify('screen_action', '{"action": "long_click"}'),
        ToolRisk.high,
      );
    });

    test('classify_screenActionScroll_returnsHigh', () {
      expect(
        classifier.classify('screen_action', '{"action": "scroll"}'),
        ToolRisk.high,
      );
    });

    test('classify_screenActionSwipe_returnsHigh', () {
      expect(
        classifier.classify('screen_action', '{"action": "swipe"}'),
        ToolRisk.high,
      );
    });

    test('classify_allSafeToolsWithAnyArgs_remainSafe', () {
      for (final tool in [
        'calculator',
        'timer',
        'device_info',
        'notepad',
        'screen_reader',
        'screen_find',
        'notification_reader',
        'contact_search',
      ]) {
        expect(
          classifier.classify(tool, '{"random": "args"}'),
          ToolRisk.safe,
          reason: '$tool should be safe regardless of args',
        );
      }
    });
  });
}

File: lib/src/__tests__/unit/agent/risk-classifier.test.ts
import { describe, it, expect } from 'vitest';
import { RiskClassifier } from '../../../agent/risk-classifier';

describe('RiskClassifier', () => {
  const classifier = new RiskClassifier();

  describe('safe tools', () => {
    const safeTools = ['calculator', 'timer', 'device_info', 'notepad', 'screen_reader', 'screen_find', 'notification_reader', 'contact_search'];

    for (const tool of safeTools) {
      it(`classifies ${tool} as safe`, () => {
        expect(classifier.classify(tool, '{}')).toBe('safe');
      });
    }

    it('classifies safe tools regardless of args', () => {
      expect(classifier.classify('calculator', '{"expression": "2+3"}')).toBe('safe');
      expect(classifier.classify('notepad', '{"action": "save"}')).toBe('safe');
    });
  });

  describe('app_launcher', () => {
    it('classifies open_settings as low', () => {
      expect(classifier.classify('app_launcher', '{"action": "open_settings"}')).toBe('low');
    });

    it('classifies list_apps as low', () => {
      expect(classifier.classify('app_launcher', '{"action": "list_apps"}')).toBe('low');
    });

    it('classifies open_app as high', () => {
      expect(classifier.classify('app_launcher', '{"action": "open_app"}')).toBe('high');
    });

    it('classifies open_url as high', () => {
      expect(classifier.classify('app_launcher', '{"action": "open_url"}')).toBe('high');
    });

    it('classifies unknown action as low', () => {
      expect(classifier.classify('app_launcher', '{"action": "search"}')).toBe('low');
    });

    it('classifies empty action as low', () => {
      expect(classifier.classify('app_launcher', '{}')).toBe('low');
    });

    it('is case-insensitive for action', () => {
      expect(classifier.classify('app_launcher', '{"action": "OPEN_APP"}')).toBe('high');
      expect(classifier.classify('app_launcher', '{"action": "Open_Settings"}')).toBe('low');
    });
  });

  describe('screen_action', () => {
    it('classifies global as low', () => {
      expect(classifier.classify('screen_action', '{"action": "global"}')).toBe('low');
    });

    it('classifies type with non-sensitive content as low', () => {
      expect(classifier.classify('screen_action', '{"action": "type", "content": "hello world"}')).toBe('low');
    });

    it('classifies type with password as critical', () => {
      expect(classifier.classify('screen_action', '{"action": "type", "content": "enter password123"}')).toBe('critical');
    });

    it('classifies type with PIN as critical', () => {
      expect(classifier.classify('screen_action', '{"action": "type", "content": "enter pin code"}')).toBe('critical');
    });

    it('classifies type with passcode as critical', () => {
      expect(classifier.classify('screen_action', '{"action": "type", "content": "passcode"}')).toBe('critical');
    });

    it('classifies type with SSN as critical', () => {
      expect(classifier.classify('screen_action', '{"action": "type", "content": "SSN number"}')).toBe('critical');
    });

    it('classifies type with credit card as critical', () => {
      expect(classifier.classify('screen_action', '{"action": "type", "content": "credit card number"}')).toBe('critical');
    });

    it('classifies type with CVV as critical', () => {
      expect(classifier.classify('screen_action', '{"action": "type", "content": "CVV"}')).toBe('critical');
    });

    it('classifies type with OTP as critical', () => {
      expect(classifier.classify('screen_action', '{"action": "type", "content": "OTP code"}')).toBe('critical');
    });

    it('classifies tap as low', () => {
      expect(classifier.classify('screen_action', '{"action": "tap"}')).toBe('low');
    });

    it('classifies scroll as low', () => {
      expect(classifier.classify('screen_action', '{"action": "scroll"}')).toBe('low');
    });
  });

  describe('sms_sender', () => {
    it('classifies send as critical', () => {
      expect(classifier.classify('sms_sender', '{"action": "send"}')).toBe('critical');
    });

    it('classifies read as high', () => {
      expect(classifier.classify('sms_sender', '{"action": "read"}')).toBe('high');
    });

    it('classifies empty action as high', () => {
      expect(classifier.classify('sms_sender', '{}')).toBe('high');
    });
  });

  describe('phone_caller', () => {
    it('classifies call as critical', () => {
      expect(classifier.classify('phone_caller', '{"action": "call"}')).toBe('critical');
    });

    it('classifies dial as high', () => {
      expect(classifier.classify('phone_caller', '{"action": "dial"}')).toBe('high');
    });

    it('classifies empty action as high', () => {
      expect(classifier.classify('phone_caller', '{}')).toBe('high');
    });
  });

  describe('unknown tool', () => {
    it('classifies unknown tool as high', () => {
      expect(classifier.classify('unknown_tool', '{}')).toBe('high');
    });

    it('classifies empty tool name as high', () => {
      expect(classifier.classify('', '{}')).toBe('high');
    });
  });

  describe('edge cases', () => {
    it('handles empty args string', () => {
      expect(classifier.classify('calculator', '')).toBe('safe');
    });

    it('handles invalid JSON', () => {
      expect(() => classifier.classify('calculator', 'not json')).toThrow();
    });

    it('handles case-insensitive action for screen_action', () => {
      expect(classifier.classify('screen_action', '{"action": "Type", "content": "Password"}')).toBe('critical');
    });
  });
});
